<robot name="my_robot">
  <link name="fito_world"></link>
  <joint name="base_joint" type="fixed">
    <parent link="fito_world" />
    <child link="base_link" />
    <origin xyz="0 0 0" rpy="0 0 0" />
  </joint>
<link name="base_footprint"></link>

<joint name="base_footprint_joint" type="fixed">
    <parent link="fito_world" />
    <child link="base_footprint" />
    <origin xyz="0 0 0" rpy="0 0 0" />
  </joint>


  <link name="base_link">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <box size="2.11 1.28 0.12" />
      </geometry>
      <material name="base_link-material">
        <color rgba="0.02732089163382382 0.5394794890033748 0.20863687013464577 1" />
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <box size="2.11 1.28 0.12" />
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="400" />
      <inertia ixx="55.1" ixy="0" ixz="0" iyy="148.9" iyz="0" izz="203" />
    </inertial>
  </link>
  <joint name="base_link_to_wheeling_mech" type="revolute">
    <parent link="base_link" />
    <child link="wheeling_mech" />
    <origin xyz="-1.051211695797214 0 0" rpy="0 0 0" />
    <axis xyz="0 0 1"/>
    <limit effort="3000.0" lower="-0.4" upper="0.4" velocity="0.5"/>  
    <dynamics damping="0.1" friction="0.5"/>
  </joint>
  <link name="wheeling_mech">
    <visual>
      <origin xyz="-0.3 0 0" rpy="0 0 0" />
      <geometry>
        <box size="0.64 0.1 0.12" />
      </geometry>
      <material name="wheeling_mech-material">
        <color rgba="0.3515325994898463 0.07036009568874305 0.5775804404214573 1" />
      </material>
    </visual>
    <collision>
      <origin xyz="-0.3 0 0" rpy="0 0 0" />
      <geometry>
        <box size="0.64 0.1 0.12" />
      </geometry>
    </collision>

    <inertial>
      <origin xyz="-0.3 0 0" rpy="0 0 0" />
      <mass value="50" />
      <inertia ixx="0.102" ixy="0" ixz="0" iyy="1.763" iyz="0" izz="1.733" />
    </inertial>
  </link>
  <joint name="wheeling_mech_to_front_wheels_base" type="fixed">
    <parent link="wheeling_mech" />
    <child link="front_wheels_base" />
    <origin xyz="-0.5850418798406349 0 0" rpy="0 0 0" />
  </joint>
  <link name="front_wheels_base">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <box size="0.082 1.28 0.12" />
      </geometry>
      <material name="front_wheels_base-material">
        <color rgba="0.5028864580233624 0.025186859622305935 0.41254261347374327 1" />
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <box size="0.082 1.28 0.12" />
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="50" />
      <inertia ixx="6.893" ixy="0" ixz="0" iyy="0.102" iyz="0" izz="6.834" />
    </inertial>
  </link>
  <joint name="front_wheels_base_to_front_right_base" type="fixed">
    <parent link="front_wheels_base" />
    <child link="front_right_base" />
    <origin xyz="0 0.63 -0.36" rpy="0 0 0" />
  </joint>
  <link name="front_right_base">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <box size="0.175 0.04 0.83" />
      </geometry>
      <material name="front_right_base-material">
        <color rgba="0.01680737574872402 0.010960094003125918 0.21952619971859377 1" />
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <box size="0.175 0.04 0.83" />
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="15" />
      <inertia ixx="0.86" ixy="0" ixz="0" iyy="0.89" iyz="0" izz="0.04" />
    </inertial>
  </link>
  <joint name="front_right_base_to_front_right_wheel" type="continuous">
    <parent link="front_right_base" />
    <child link="front_right_wheel" />
    <origin xyz="0 0.09 -0.41" rpy="1.5707963267948963 0 0" />
    <axis xyz="0 0 1"/>
  </joint>
  <link name="front_right_wheel">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <cylinder radius="0.37" length="0.16" />
      </geometry>
      <material name="front_right_wheel-material">
        <color rgba="0.013702083043526807 0.16826940017946088 0.009134058699157796 1" />
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <cylinder radius="0.37" length="0.16" />
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="60" />
      <inertia ixx="2.18" ixy="0" ixz="0" iyy="2.18" iyz="0" izz="4.107" />
    </inertial>
  </link>
  <joint name="front_wheels_base_to_front_left_base" type="fixed">
    <parent link="front_wheels_base" />
    <child link="front_left_base" />
    <origin xyz="0 -0.63 -0.36" rpy="0 0 0" />
  </joint>
  <link name="front_left_base">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <box size="0.175 0.04 0.83" />
      </geometry>
      <material name="front_left_base-material">
        <color rgba="0.01680737574872402 0.010960094003125918 0.21952619971859377 1" />
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <box size="0.175 0.04 0.83" />
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="15" />
      <inertia ixx="0.86" ixy="0" ixz="0" iyy="0.89" iyz="0" izz="0.04" />
    </inertial>
  </link>
  <joint name="front_left_base_to_front_left_wheel" type="continuous">
    <parent link="front_left_base" />
    <child link="front_left_wheel" />
    <origin xyz="0 -0.09 -0.41" rpy="1.5707963267948963 0 0" />
    <axis xyz="0 0 1"/>
  </joint>
  <link name="front_left_wheel">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <cylinder radius="0.37" length="0.16" />
      </geometry>
      <material name="front_left_wheel-material">
        <color rgba="0.11443537381770343 0.35640014413537763 0.11953842797895521 1" />
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <cylinder radius="0.37" length="0.16" />
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="60" />
      <inertia ixx="2.18" ixy="0" ixz="0" iyy="2.18" iyz="0" izz="4.107" />
    </inertial>
  </link>
  <joint name="base_link_to_back_left_base" type="fixed">
    <parent link="base_link" />
    <child link="back_left_base" />
    <origin xyz="0.8122615873599781 -0.62 -0.3556037303067785" rpy="0 0 0" />
  </joint>
  <link name="back_left_base">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <box size="0.175 0.04 0.83" />
      </geometry>
      <material name="back_left_base-material">
        <color rgba="0.8631572134510892 0.5457244613615395 0.04817182422013895 1" />
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <box size="0.175 0.04 0.83" />
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="15" />
      <inertia ixx="0.86" ixy="0" ixz="0" iyy="0.89" iyz="0" izz="0.04" />
    </inertial>
  </link>
  <joint name="back_left_base_to_back_left_wheel" type="continuous">
    <parent link="back_left_base" />
    <child link="back_left_wheel" />
    <origin xyz="0 -0.09 -0.41" rpy="1.5707963267948963 0 0" />
    <axis xyz="0 0 1"/>
  </joint>
  <link name="back_left_wheel">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <cylinder radius="0.37" length="0.16" />
      </geometry>
      <material name="back_left_wheel-material">
        <color rgba="0.2663556047920505 0.4396571738310091 0.006048833020386069 1" />
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <cylinder radius="0.37" length="0.16" />
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="60" />
      <inertia ixx="2.18" ixy="0" ixz="0" iyy="2.18" iyz="0" izz="4.107" />
    </inertial>
  </link>
  <joint name="base_link_to_back_right_base" type="fixed">
    <parent link="base_link" />
    <child link="back_right_base" />
    <origin xyz="0.8122615873599781 0.62 -0.36" rpy="0 0 0" />
  </joint>
  <link name="back_right_base">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <box size="0.175 0.04 0.83" />
      </geometry>
      <material name="back_right_base-material">
        <color rgba="0.8631572134510892 0.5457244613615395 0.04817182422013895 1" />
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <box size="0.175 0.04 0.83" />
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="15" />
      <inertia ixx="0.86" ixy="0" ixz="0" iyy="0.89" iyz="0" izz="0.04" />
    </inertial>
  </link>
  <joint name="back_right_base_to_back_right_wheel" type="continuous">
    <parent link="back_right_base" />
    <child link="back_right_wheel" />
    <origin xyz="0 0.09 -0.41" rpy="1.5707963267948963 0 0" />
    <axis xyz="0 0 1"/>
  </joint>
  <link name="back_right_wheel">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <cylinder radius="0.37" length="0.16" />
      </geometry>
      <material name="back_right_wheel-material">
        <color rgba="0.9301108583738498 0.3515325994898463 0.8307698767709715 1" />
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <cylinder radius="0.37" length="0.16" />
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="60" />
      <inertia ixx="2.18" ixy="0" ixz="0" iyy="2.18" iyz="0" izz="4.107" />
    </inertial>
  </link>
  <joint name="base_link_to_back_load" type="fixed">
    <parent link="base_link" />
    <child link="back_load" />
    <origin xyz="0.46646768743264955 0 0.2753277941857757" rpy="0 0 0" />
  </joint>
  <link name="back_load">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <box size="0.5 1 0.5" />
      </geometry>
      <material name="back_load-material">
        <color rgba="0.5457244613615395 0.3005437944049895 0.05612849004241121 1" />
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <box size="0.5 1 0.5" />
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="50" />
      <inertia ixx="5.208" ixy="0" ixz="0" iyy="2.083" iyz="0" izz="5.208" />
    </inertial>
  </link>
  <joint name="base_link_to_front_load" type="fixed">
    <parent link="base_link" />
    <child link="front_load" />
    <origin xyz="-0.32806438331095766 0 0.2753277941857757" rpy="0 0 0" />
  </joint>
  <link name="front_load">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <box size="0.5 1 0.5" />
      </geometry>
      <material name="front_load-material">
        <color rgba="0.5457244613615395 0.3005437944049895 0.05612849004241121 1" />
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <box size="0.5 1 0.5" />
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="60" />
      <inertia ixx="6.25" ixy="0" ixz="0" iyy="2.5" iyz="0" izz="6.25" />
    </inertial>
  </link>
  <joint name="base_link_to_lidar_base" type="fixed">
    <parent link="base_link" />
    <child link="lidar_base" />
    <origin xyz="-0.92 0.002145114131801129 0.6811955493557079" rpy="0 0 0" />
  </joint>
  <link name="lidar_base">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <cylinder radius="0.01" length="1.2" />
      </geometry>
      <material name="lidar_base-material">
        <color rgba="0.06662593863608139 0.5906188409113381 0.008023192982520563 1" />
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <cylinder radius="0.01" length="1.2" />
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="1" />
      <inertia ixx="0.3333333333333333" ixy="0" ixz="0" iyy="0.5" iyz="0" izz="0.3333333333333333" />
    </inertial>
  </link>
  <joint name="lidar_base_to_lidar" type="fixed">
    <parent link="lidar_base" />
    <child link="lidar" />
    <origin xyz="0 0 0.6172694527420378" rpy="0 0 0" />
  </joint>
  <link name="lidar">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <sphere radius="0.05" />
      </geometry>
      <material name="lidar-material">
        <color rgba="0.1912016827303171 0.03954623527052923 0.85499260812105 1" />
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <sphere radius="0.05" />
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="1" />
      <inertia ixx="0.4" ixy="0" ixz="0" iyy="0.4" iyz="0" izz="0.4" />
    </inertial>


      </link>
  <gazebo reference="lidar">
    <sensor name="gpu_lidar" type="gpu_lidar">
      <update_rate>10</update_rate>
      <topic>scan</topic>
      <gz_frame_id>lidar</gz_frame_id>
      <lidar>
        <scan>
          <horizontal>
            <samples>720</samples>
            <!--(max_angle-min_angle)/samples * resolution -->
            <resolution>1</resolution>
            <min_angle>-3.14156</min_angle>
            <max_angle>3.14156</max_angle>
          </horizontal>
          <!-- Dirty hack for fake lidar detections with ogre 1 rendering in VM -->
          <!-- <vertical>
              <samples>3</samples>
              <min_angle>-0.001</min_angle>
              <max_angle>0.001</max_angle>
          </vertical> -->
        </scan>
        <range>
          <min>0.05</min>
          <max>10.0</max>
          <resolution>0.01</resolution>
        </range>
        <noise>
            <type>gaussian</type>
            <mean>0.0</mean>
            <stddev>0.01</stddev>
        </noise>
        <frame_id>lidar</frame_id>
      </lidar>
      <always_on>1</always_on>
      <visualize>true</visualize>
    </sensor>
  </gazebo>

    <ros2_control name="GazeboSystem" type="system">
      <hardware>
        <plugin>gz_ros2_control/GazeboSimSystem</plugin>
      </hardware>


      <joint name="back_right_base_to_back_right_wheel">
        <command_interface name="velocity">
          <param name="min">-10</param>
          <param name="max">10</param>
        </command_interface>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
      </joint>
     
      <joint name="back_left_base_to_back_left_wheel">
        <command_interface name="velocity">
          <param name="min">-10</param>
          <param name="max">10</param>
        </command_interface>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
      </joint>

      <joint name="front_right_base_to_front_right_wheel">
        <command_interface name="velocity">
          <param name="min">-10</param>
          <param name="max">10</param>
        </command_interface>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
      </joint>
     
      <joint name="front_left_base_to_front_left_wheel">
        <command_interface name="velocity">
          <param name="min">-10</param>
          <param name="max">10</param>
        </command_interface>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
      </joint>
      
      

      <!-- Steering mechanisms (position control) -->
      <joint name="base_link_to_wheeling_mech">
        <command_interface name="position">
          <param name="min">-0.4</param>
          <param name="max">0.4</param>
        </command_interface>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
      </joint>
      
      
    </ros2_control>
  

 
    <gazebo>
      <plugin name="gz_ros2_control::GazeboSimROS2ControlPlugin" filename="libgz_ros2_control-system.so">
        <parameters>/home/alexey/ros2_ws/src/my_robot/config/new_controllers.yaml</parameters>
      </plugin>
   
    </gazebo>
    
 
    <gazebo reference="base_link">
      <material>Gazebo/Grey</material>
    </gazebo>
 
</robot>